### ISSUE TITLE
Can't verify webhook sandbox events (not simulator) 

### ISSUE THREAD

**Issue by trixyz** [paypal/PayPal-node-SDK#294]
### General information

* SDK/Library version: 1.7.1
* Environment: sandbox
* `PayPal-Debug-ID` values: -
* Language, language version, and OS: node 8.2.1, express 

### Issue description
i'm using paypal.notification.webhookEvent.verify method and always got failure.

here is example of latest event:

Headers object i pass as headers arg:
`{ host: '93.85.94.98',
  connection: 'close',
  'content-length': '1584',
  accept: '*/*',
  'paypal-transmission-id': '47a57790-7d0e-11e7-ac1e-6b62a8a99ac4',
  'paypal-transmission-time': '2017-08-09T14:23:14Z',
  'paypal-transmission-sig': 'ZA4VaHAn4/+/F0+RqlL+QjEbx50iLxCiNhdt1XfeBUwJtxCPU1OkDOPJF2vHq1QPuYF0XEk5TomrAlgkJc4+rbee8Vcm9CVftAQN2UGTYaRe3osGxFFFADgefNHXV/W8DkvE6xo3e30kqSASYfaqyxckjxUl2DnNFsX1mIHI8Z06O+s7PSXFN3sZbp4JyfJvRY+6uyT0VokNa7jCKNSBRhwoNq0vtL7L+VJ0jfPcUfY72JHb8HIhRaBwUwR89mB/ejPTriZQKVOWh2HrqikLvaQWm/eUEduE54RIIrgEf6/ruqYEEd2iTOL9CMZp5AqhkNrT0TIlSwHgJgGyv1wzXg==',
  'paypal-auth-version': 'v2',
  'paypal-cert-url': 'https://api.sandbox.paypal.com/v1/notifications/certs/CERT-360caa42-fca2a594-aecacc47',
  'paypal-auth-algo': 'SHA256withRSA',
  'content-type': 'application/json',
  'user-agent': 'PayPal/AUHD-211.0-33754056',
  'correlation-id': '3d92e1964b3f9' }`

request body: 
`{"id":"WH-5CK508006B546110N-3E048491YS948403K","event_version":"1.0","create_time":"2017-08-09T14:23:09.883Z","resource_type":"invoices","event_type":"INVOICING.INVOICE.PAID","summary":"An invoice has been paid","resource":{"invoice":{"id":"INV2-8D3A-CPGB-9BZV-FM6H","number":"0064","template_id":"TEMP-46M523821A2033342","status":"PAYMENT_PENDING","merchant_info":{"email":"support-facilitator@secondwarehouse.com"},"billing_info":[{"email":"support-buyer@secondwarehouse.com"}],"shipping_info":{"first_name":"dear customer"},"items":[{"name":"Commission fee for...","quantity":1.0,"unit_price":{"currency":"EUR","value":"15.00"}}],"invoice_date":"2017-08-09 PDT","tax_calculated_after_discount":false,"tax_inclusive":false,"total_amount":{"currency":"EUR","value":"15.00"},"payments":[{"type":"PAYPAL","transaction_id":"59722583FG433445B","transaction_type":"SALE","method":"PAYPAL","date":"2017-08-09 07:22:44 PDT","amount":{"currency":"EUR","value":"15.00"}}],"metadata":{"created_date":"2017-08-09 07:21:12 PDT","last_updated_date":"2017-08-09 07:22:24 PDT","first_sent_date":"2017-08-09 07:21:13 PDT","last_sent_date":"2017-08-09 07:21:13 PDT"},"links":[{"rel":"self","href":"https://api.sandbox.paypal.com/v1/invoicing/invoices/INV2-8D3A-CPGB-9BZV-FM6H","method":"GET"}]}},"links":[{"href":"https://api.sandbox.paypal.com/v1/notifications/webhooks-events/WH-5CK508006B546110N-3E048491YS948403K","rel":"self","method":"GET"},{"href":"https://api.sandbox.paypal.com/v1/notifications/webhooks-events/WH-5CK508006B546110N-3E048491YS948403K/resend","rel":"resend","method":"POST"}]}
`
i tried to pass them as object after body is parsed by bodyParser and also tried to pass simply as string read directly without bodyParser like:
`req.on("data", (chunk) => (rawData += chunk))`

here is my webhook id:
`02U57613FJ0952018`

**Comment by pp-randy** [comment]
It looks like validation is passing, but a possible down stream error.  

Is the validation API just returning validation failed or another error?

**Comment by pp-randy** [comment]
Can you provide the paypal-debug-id from validation failure?

**Comment by trixyz** [comment]
i just got { verification_status: 'FAILURE', httpStatusCode: 200 } in response

**Comment by pp-randy** [comment]
Can you toss me the headers?

**Comment by trixyz** [comment]
it is from verify response i get just at the moment
`{ date: 'Wed, 09 Aug 2017 15:39:35 GMT',
  server: 'Apache',
  'paypal-debug-id': '7087022e7df83, 7087022e7df83',
  'content-language': '*',
  connection: 'close',
  http_x_pp_az_locator: 'sandbox.slc',
  'set-cookie': 
   [ 'X-PP-SILOVER=name%3DSANDBOX3.API.1%26silo_version%3D1880%26app%3Dapiplatformproxyserv%26TIME%3D3073149785%26HTTP_X_PP_AZ_LOCATOR%3Dsandbox.slc; Expires=Wed, 09 Aug 2017 16:09:35 GMT; domain=.paypal.com; path=/; Secure; HttpOnly',
     'X-PP-SILOVER=; Expires=Thu, 01 Jan 1970 00:00:01 GMT' ],
  vary: 'Authorization',
  'content-length': '33',
  'content-type': 'application/json' }
`

**Comment by pp-randy** [comment]
I was not able to see the request body; however, this is also a common cause of the error https://github.com/paypal/PayPal-REST-API-issues/issues/18 and the change can happen when your are sending data back for validation due to JSON parsing and the bug with quantity being changed.

**Comment by trixyz** [comment]
Thanks, so is there some workaround to pass this issue to be able verify webhooks for invoices?

**Comment by pp-randy** [comment]
You need to accept the raw request body, but also send a raw body back.  Due to how JavaScript JSON parsing works. 

The SDK won't work for invoicing only because of the JSON parsing here https://github.com/paypal/PayPal-node-SDK/blob/master/lib/resources/Notification.js#L119.  

I've had success using https://www.npmjs.com/package/httpreq and the `body` argument. Due to the JSON parsing issues I manually create the request.
```
let webhookObjJSON = '{';
	webhookObjJSON += '"auth_algo": "' + req.header('PAYPAL-AUTH-ALGO') + '"';
        webhookObjJSON += ',"cert_url": "' + req.header('PAYPAL-CERT-URL')+ '"';
        webhookObjJSON += ',"transmission_id": "' + req.header('PAYPAL-TRANSMISSION-ID')+ '"';
        webhookObjJSON += ',"transmission_sig": "' + req.header('PAYPAL-TRANSMISSION-SIG')+ '"';
        webhookObjJSON += ',"transmission_time": "' + req.header('PAYPAL-TRANSMISSION-TIME')+ '"';
        webhookObjJSON += ',"webhook_id": "<INSERT WEBHOOK ID>"',
        webhookObjJSON += ',"webhook_event": ' + req.body;
	webhookObjJSON += '}';
let validationReqOptions = {
	headers: {
		'Authorization': 'Bearer ' + JSON.parse(tokenres.body).access_token,
		'content-type': 'application/json'
	},
	body: webhookObjJSON
};

httpreq.post('https://api.sandbox.paypal.com/v1/notifications/verify-webhook-signature', validationReqOptions, (err, ppres) => {
  
```

**Comment by xiaocc2012** [comment]
thanks @pp-randy ，i solve the problem by manually create the request

**Comment by christineurban** [comment]
Manually creating the request exactly as shown above with httpreq does not work for me, I get this response: 

```
{"name":"VALIDATION_ERROR","message":"Invalid request - see details","debug_id":"83892b4f31e1","details":[{"location":"body","issue":"Cannot deserialize instance of `java.util.LinkedHashMap` out of START_ARRAY token line: 1, column: 643"}],"links":[]}
```

And if I stringify `req.body` I get 
```
{ verification_status: 'FAILURE'}
```

parsed JSON body
```
{
	"auth_algo": "SHA256withRSA",
	"cert_url": "https://api.paypal.com/v1/notifications/certs/CERT-360caa42-fca2a594-ed3bf26f",
	"transmission_id": "1ea52e80-3a4f-11e9-9aa7-3195749595e0",
	"transmission_sig": "qWafWSSdXieWxyvD4my38t1npcFwqdUqxtDyStHKa4i1aeq/eeyi2p6f/JnlD+ZyVadbA/FvoZVBDs6N7WvLEHjB7uX26t43XVj0EGYE5iJCovp2K+5zJxp4nMsPaF05EMJOFVZSdIOw5u3kZDOKFFr//jX+DJm+cmkjRI2bf/SnFzvJKXY0xsnl+4Pv0tE6jOddClZjdBug6mHd+3yFkBLMN55vVF0mcB7e/tdvzJsCufVithVxpN7d/h4Xxr2sgp+OnpyegmGM2yTEX3zZru+ya6Q9dD64w2jMIe2P/nWoCdtrxMXmsogEQUxgyCnfGbFkfm5Ab0KipS+TDTpSzA==",
	"transmission_time": "2019-02-27T05:18:30Z",
	"webhook_id": "2AB5537978016141J",
	"webhook_event": {
		"id": "WH-19973937YW279670F-02S63370HL636500Y",
		"event_version": "1.0",
		"create_time": "2016-04-28T11:29:31Z",
		"resource_type": "Agreement",
		"event_type": "BILLING.SUBSCRIPTION.CREATED",
		"summary": "A billing subscription was created",
		"resource": {
			"id": "I-PE7JWXKGVN0R",
			"shipping_address": {
				"recipient_name": "Cool Buyer",
				"line1": "3rd st",
				"line2": "cool",
				"city": "San Jose",
				"state": "CA",
				"postal_code": "95112",
				"country_code": "US"
			},
			"plan": {
				"curr_code": "USD",
				"links": [],
				"payment_definitions": [{
					"type": "TRIAL",
					"frequency": "Month",
					"frequency_interval": "1",
					"amount": {
						"value": "5.00"
					},
					"cycles": "5",
					"charge_models": [{
						"type": "TAX",
						"amount": {
							"value": "1.00"
						}
					}, {
						"type": "SHIPPING",
						"amount": {
							"value": "1.00"
						}
					}]
				}, {
					"type": "REGULAR",
					"frequency": "Month",
					"frequency_interval": "1",
					"amount": {
						"value": "10.00"
					},
					"cycles": "15",
					"charge_models": [{
						"type": "TAX",
						"amount": {
							"value": "2.00"
						}
					}, {
						"type": "SHIPPING",
						"amount": {
							"value": "1.00"
						}
					}]
				}],
				"merchant_preferences": {
					"setup_fee": {
						"value": "0.00"
					},
					"auto_bill_amount": "YES",
					"max_fail_attempts": "21"
				}
			},
			"payer": {
				"payment_method": "paypal",
				"status": "verified",
				"payer_info": {
					"email": "coolbuyer@example.com",
					"first_name": "Cool",
					"last_name": "Buyer",
					"payer_id": "XLHKRXRA4H7QY",
					"shipping_address": {
						"recipient_name": "Cool Buyer",
						"line1": "3rd st",
						"line2": "cool",
						"city": "San Jose",
						"state": "CA",
						"postal_code": "95112",
						"country_code": "US"
					}
				}
			},
			"agreement_details": {
				"outstanding_balance": {
					"value": "0.00"
				},
				"num_cycles_remaining": "5",
				"num_cycles_completed": "0",
				"final_payment_due_date": "2017-11-30T10:00:00Z",
				"failed_payment_count": "0"
			},
			"description": "desc",
			"state": "Pending",
			"links": [{
				"href": "https://api.paypal.com/v1/payments/billing-agreements/I-PE7JWXKGVN0R",
				"rel": "self",
				"method": "GET"
			}],
			"start_date": "2016-04-30T07:00:00Z"
		},
		"links": [{
			"href": "https://api.paypal.com/v1/notifications/webhooks-events/WH-19973937YW279670F-02S63370HL636500Y",
			"rel": "self",
			"method": "GET"
		}, {
			"href": "https://api.paypal.com/v1/notifications/webhooks-events/WH-19973937YW279670F-02S63370HL636500Y/resend",
			"rel": "resend",
			"method": "POST"
		}]
	}
}
```

**Comment by xiaocc2012** [comment]
I suffer this problem nearly one year ago.
The error message you received seems that, your request body is error
format, not a serializered json as string.

In order to send a http request, the request body has two method to
construct.
1. construct a json object, use its serializered function to get a string.
2. manually connect chars to construct a string, like '{' + ' "a" ' + ':' +
' "b" ' + '}', but you must notice the double quotes, without that, it's
a grammatical error.

At last, the sdk api is designed very strange, i think that the `webhook_event`
field in verify api, should not an object, but a string.
Because its create server and verify server often not the same, even the
coding language are not same.
Json object serializered to string, don't have a standard specification,
you can't suppose the fields keys ordered from small to big.

Christine Urban <notifications@github.com> 于2019年2月27日周三 下午3:02写道：

**Comment by Gaen** [comment]
Tried solution from @randstraw, worked indeed.

But then I went further and tried sending `body` as `JSON.stringify(JSON.parse(webhookObjJSON))` and this was enough to break it. So, I compared two request bodies (valid and invalid) and found exactly one difference.

Valid:
`"quantity": 1.0,`

Invalid:
`"quantity": 1,`

Looks like PayPal is sending the number as `1.0` and expects it the same way for the verification to succeed while `JSON.parse`/`JSON.stringify` transforms it into `1` and there you go.
